### PR TITLE
Modify search specimen logic

### PR DESCRIPTION
**Pull request by Gbarra9**
The following Pull request (PR) addresses [Add Deviations to Box Manifests - #609](https://github.com/episphere/connect/issues/609) :

The PR is linked to the [Search specimen bptl click manifest #527 PR](https://github.com/episphere/biospecimen/pull/527)
- Reference [searchSpecimenByRequestedSite](https://github.com/episphere/biospecimen/pull/527/files#diff-b1ad84bc7117076c54df7562cdfba01557615bde3c5fe26d1c247789d695ba24R866-R889) in Biospecimen Repo

**Updates to the code:** 
- created [convertSiteLoginToNumber function](https://github.com/episphere/connectFaas/pull/351/files#diff-23fab349744c189c5b655022c3edc921526e74945e34cdd65db82880911119d8R677-R683) to convert login site's `string type` to `number type` and will return a `number value` if found or `undefined` otherwise
- created a [fieldToConceptIdMapping.js](https://github.com/episphere/connectFaas/pull/351/files#diff-0d8ec4ec095b077565be32aa751d82e00f80837c8c53aeb2d26e37ffe2504481R1-R39) file to reference `concept Ids` and replaced concept Ids with meaningful names
- modified [searchSpecimen](https://github.com/episphere/connectFaas/pull/351/files#diff-c359139ac64daf8f5ece96e5f36b7847c5c63aae960815984bf13d414b8b76b6R210) to allow searchSpecimen's `searchShipments` function to take in a requested login site's number value
- refactored `searchShipments` function to reference `fieldToConceptIdMapping.js` and replaced existing `conceptIds` with `fieldMapping` reference
- updated package.json's devDependencies `@google-cloud/functions-framework` from "^1.7.1" to "^3.2.0"

**Changes from PR comments and more:**
- refactored [searchShipments](https://github.com/episphere/connectFaas/pull/351/files#diff-be91e66430aa4dd2f61668d1bd17d3cb449f57947e13bfa2f6bc087919967282R1000-R1030) by replacing JS array`filter` method and JS array `map` method with `for... loop`
- added [swapObjKeysAndValues](https://github.com/episphere/connectFaas/pull/351/files#diff-23fab349744c189c5b655022c3edc921526e74945e34cdd65db82880911119d8R685-R693)  function to correctly reference [tubesBagCids](https://github.com/episphere/connectFaas/pull/351/files#diff-0d8ec4ec095b077565be32aa751d82e00f80837c8c53aeb2d26e37ffe2504481R22-R38)
- added "no" variable to [obj assignment destructuring](https://github.com/episphere/connectFaas/pull/351/commits/ccd44e0d8ae58abcf65341631f27c6cc86d1d2ab#diff-be91e66430aa4dd2f61668d1bd17d3cb449f57947e13bfa2f6bc087919967282R1001) to work properly in searchShipments function's conditional logic

### CODE REVIEW

File: utils/firestore.js [hunk 1]
```diff
@@ -996,61 +998,34 @@ const searchSpecimen = async (masterSpecimenId, siteCode) => {
...
-                    if(currJSON.hasOwnProperty('258745303')){
-                        if(currJSON['258745303'] == '104430631'){
+            let isFound = false;
+
+            for (let key of keys) {
+                if (conversionTubesBags[key]) {
+                    let currJSON = data[key];
+
+                    if (currJSON[reportMissingTube]) {
+                        if (currJSON[reportMissingTube] === no) {
```
Comment (we-ai) [inline utils/firestore.js:1015]: `no` is no defined.

File: utils/biospecimen.js [hunk 2]
```diff
@@ -221,7 +221,8 @@ const biospecimenAPIs = async (req, res) => {
         }
         else {
             const { searchShipments } = require('./firestore');
-            const response = await searchShipments(siteCode);
+            const requestedSite = convertSiteLoginToNumber(req?.query?.requestedSite);
+            const response = requestedSite ? await searchShipments(requestedSite) : await searchShipments(siteCode);
```
Comment (we-ai) [inline utils/biospecimen.js:225]: Are these changes needed here? Will this enable cross-site search?

File: utils/biospecimen.js [hunk 3]
```diff
@@ -221,7 +221,8 @@ const biospecimenAPIs = async (req, res) => {
         }
         else {
             const { searchShipments } = require('./firestore');
-            const response = await searchShipments(siteCode);
+            const requestedSite = convertSiteLoginToNumber(req?.query?.requestedSite);
```
Comment (we-ai) [inline utils/biospecimen.js:224]:
```suggestion
            const requestedSite = convertSiteLoginToNumber(req.query?.requestedSite);
```

File: utils/fieldToConceptIdMapping.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        Kaiser_Permanente_Georgia: 327912200,
+        Sanford_Health: 657167265,
+        University_of_Chicago_Medicine: 809703864,
+        Kaiser_Permanente_Hawaii: 300267574,
+        Marshfield_Clinic: 303349821,
+        Kaiser_Permanente_Northwest: 452412599,
+        Kaiser_Permanente_Colorado: 125001209,
+        HealthPartners_Research_Clinic: 531629870,
+        Henry_Ford_Health_System: 548392715,
+        National_Institutes_of_Health: 13,
```
Comment (we-ai) [inline utils/fieldToConceptIdMapping.js:13]: Use camelCase or snake_case.

File: utils/fieldToConceptIdMapping.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    },
+
+    yes: "353358909",
+    no: "104430631",
+    healthCareProvider: "827220437",
+    reportMissingTube:"258745303",
+    submitShipmentFlag: "145971562",
```
Comment (we-ai) [inline utils/fieldToConceptIdMapping.js:20]: These are numbers, not strings.

File: utils/firestore.js [hunk 6]
```diff
@@ -996,61 +998,34 @@ const searchSpecimen = async (masterSpecimenId, siteCode) => {
 }
 
 const searchShipments = async (siteCode) => {
-    const snapshot = await db.collection('biospecimen').where('827220437', '==', siteCode).get();
+    const { reportMissingTube, submitShipmentFlag, healthCareProvider} = fieldMapping;
+    const snapshot = await db.collection('biospecimen').where(healthCareProvider, '==', siteCode).get();
+
     if(snapshot.size !== 0){
-        //
         return snapshot.docs.filter(document => {
```
Comment (we-ai) [inline utils/firestore.js:1005]: This filter call is bulky.

File: utils/firestore.js [hunk 7]
```diff
@@ -996,61 +998,34 @@ const searchSpecimen = async (masterSpecimenId, siteCode) => {
...
                     }
                 }
             }
-            if(found == false){
-                return true;
-            }
-            else{
-                return false;
-            }
+            return (isFound === false) ? true : false;
```
Comment (we-ai) [inline utils/firestore.js:1028]:
```suggestion
            return !isFound;
```

File: utils/shared.js [hunk 8]
```diff
@@ -672,6 +674,14 @@ const cleanSurveyData = (data) => {
     return data;
 }
 
+const convertSiteLoginToNumber = (siteLogin) => {
```
Comment (we-ai) [inline utils/shared.js:677]: returns `null` is better.

File: utils/firestore.js (hunk 1 above)
Comment (Gbarra9) [reply utils/firestore.js:1018]: Thank you for catching this.

File: utils/fieldToConceptIdMapping.js (hunk 4 above)
Comment (Gbarra9) [reply utils/fieldToConceptIdMapping.js]: Thanks, I switched to camelCase

File: utils/biospecimen.js (hunk 2 above)
Comment (Gbarra9) [reply utils/biospecimen.js]: For now this will retrieve the corresponding biospecimen documents based on the login site of the [box's button clicked](https://github.com/episphere/biospecimen/pull/527/files#diff-3b1b3cdfa6b45cd7638f994ae8322d8e2b4e0e5b0b3732e282e7489a7711d5ceR212). It could be used for search functionality later on in BPTL dashboard.

File: utils/firestore.js (hunk 6 above)
Comment (Gbarra9) [reply utils/firestore.js]: Thanks this is now replaced with a `for... loop`

File: utils/firestore.js [hunk 9]
```diff
@@ -996,62 +998,36 @@ const searchSpecimen = async (masterSpecimenId, siteCode) => {
...
+    const snapshot = await db.collection('biospecimen').where(healthCareProvider, '==', siteCode).get();
+
     if(snapshot.size !== 0){
-        //
-        return snapshot.docs.filter(document => {
-            
-            let data = document.data();
+        let shipmentData = [];
+
+        for(let documents of snapshot.docs) {
```
Comment (we-ai) [inline utils/firestore.js:1009]:
```suggestion
        for(let document of snapshot.docs) {
```

File: utils/firestore.js [hunk 10]
```diff
@@ -996,62 +998,36 @@ const searchSpecimen = async (masterSpecimenId, siteCode) => {
...
+    const healthCareProvider = fieldMapping.healthCareProvider.toString();
+    const tubesBagsCidKeys = swapObjKeysAndValues(tubesBagsCids);
+    const snapshot = await db.collection('biospecimen').where(healthCareProvider, '==', siteCode).get();
+
     if(snapshot.size !== 0){
-        //
-        return snapshot.docs.filter(document => {
-            
-            let data = document.data();
+        let shipmentData = [];
```
Comment (we-ai) [inline utils/firestore.js:1007]: This line can be moved before `if` block. The `else` block is not needed. Return `shipmentData` at end of function block.

**Review by we-ai — APPROVED** [review]
Looks good. Commented on minor things.

File: utils/firestore.js (hunk 10 above)
Comment (Gbarra9) [reply utils/firestore.js:1007]: Makes sense to remove the else block after moving the `shipmentData` variable outside of the `if` block. I appreciate the suggestions.

File: utils/firestore.js (hunk 9 above)
Comment (Gbarra9) [reply utils/firestore.js:1009]: Nice catch, I should've made documents singular.
